### PR TITLE
feat: format errors

### PR DESCRIPTION
**Pull request by timcadman**
Implement two main changes:

1. Added argument `return_error` to `datashield.assign` functions, giving the option either to immediately return errors or to store them to be accessible with a call to `datashield.errors`
2. Added styling to the output returned by datashield.errors.

Note: tests written for `datashield.assign.expr`; however I was unable to create errors with either `datashield.assign.table` or `datashield.assign.resource`. If you have code which will do this tell me and I'll add tests.

### CODE REVIEW

File: R/datashield.assign.R [hunk 1]
```diff
@@ -266,7 +272,7 @@ datashield.assign.table <- function(conns, symbol, table, variables=NULL, missin
 #'   })
 #' }
 #' @export
-datashield.assign.resource <- function(conns, symbol, resource, async=TRUE, success=NULL, error=NULL) {
+datashield.assign.resource <- function(conns, symbol, resource, async=TRUE, success=NULL, error=NULL, return_errors = getOption("datashield.return_errors", TRUE)) {
```
Comment (ymarcon) [inline R/datashield.assign.R:278]: This new parameter is not documented, please add it to the roxygen2 doc above the function.

File: R/datashield.assign.R [hunk 2]
```diff
@@ -364,13 +370,19 @@ datashield.assign.resource <- function(conns, symbol, resource, async=TRUE, succ
...
+      .appendError(conns@name, conditionMessage(e))
       if (.is.callback(error)) {
-        error(conns@name, e$message)
+        error(conns@name, conditionMessage(e))
       }
     })
   }
-  .checkLastErrors()
+  if(return_errors == TRUE){
+    cli_text()
```
Comment (ymarcon) [inline R/datashield.assign.R]: Looks like a new dependency, please add it to DESCRIPTION file and declare a [@importFrom](https://roxygen2.r-lib.org/articles/namespace.html#imports) tag in the above roxygen2 documentation.

File: R/datashield.assign.R [hunk 3]
```diff
@@ -364,13 +370,19 @@ datashield.assign.resource <- function(conns, symbol, resource, async=TRUE, succ
...
+    cli_alert_warning("Error: There are some DataSHIELD errors ")
```
Comment (ymarcon) [inline R/datashield.assign.R]: What happens when the execution is not in a CLI context, like a batch script execution?

**Review by ymarcon — CHANGES_REQUESTED** [review]
Not tested yet, please fix dependencies and documentation. Also I expect the man folder to be updated. In case you do not know how to do it in Rstudio, make sure in the project options that all the boxes are checked:

image

so that documentation is generated at check time.

**Comment by timcadman** [comment]
> Not tested yet, please fix dependencies and documentation. Also I expect the man folder to be updated. In case you do not know how to do it in Rstudio, make sure in the project options that all the boxes are checked:
> 
> image
> 
> so that documentation is generated at check time.

Sorry Yannick, not sure what happened yesterday ... trying to do 17 things at once!

File: R/datashield.assign.R (hunk 3 above)
Comment (timcadman) [reply R/datashield.assign.R]: Do you mean eg calling datashield.assign.expr in a loop? Or you had something else in mind?

Comment (ymarcon) [reply R/datashield.assign.R]: No, I mean when there is no command line, like when executing code with Rscript.

**Comment by timcadman** [comment]
Hi Yannick, following your comments:
- I've now tested this with an external script (tests added to package).
- You were right to suggest this, as it highlighted an issue as currently I was returning warnings, not errors. I've tweaked further now so an error is returned and this gives the correct behaviour (throwing error when running a script).

File: .DS_Store
Comment (ymarcon) [inline .DS_Store:1]: Please add `.DS_Store` to `.gitignore` and `.Rbuildignore` (thank you Apple for the noise).

File: R/datashield.errors.R [hunk 4]
```diff
@@ -1,13 +1,62 @@
...
+#' Remove Curly Brackets from Strings
+#'
+#' This function replaces all curly brackets in the input strings with parentheses.
+#'
+#' @param errors A list of strings that may contain curly brackets.
+#' @return A list of strings with curly brackets replaced by parentheses.
+#' @importFrom purrr map
+#' @importFrom stringr str_replace_all
+.remove_curly_brackets <- function(errors) {
+  errors <- errors %>% map(~str_replace_all(.x, "\\{", "("))
```
Comment (ymarcon) [inline R/datashield.errors.R]: Try to minimize the package dependencies. I don't think you need all the fancy packages dplyr, purr and stringr for that function, just go the plain old R way:

```
# Replace all curly brackets with parentheses
output_string <- gsub("\\{", "(", input_string)
output_string <- gsub("\\}", ")", output_string)
```

**Review by ymarcon — CHANGES_REQUESTED** [review]
Please rebase to master, there are conflicts in the NAMESPACE.

File: .DS_Store
Comment (ymarcon) [reply .DS_Store:1]: Still to be ignored

Comment (ymarcon) [reply .DS_Store:1]: I meant .DS_Store file is to be removed from this PR's files

File: R/datashield.errors.R [hunk 5]
```diff
@@ -1,13 +1,72 @@
...
+
+#' Remove Curly Brackets from Strings
+#'
+#' This function replaces all curly brackets in the input strings with parentheses.
+#'
+#' @param errors A list of strings that may contain curly brackets.
+#' @return A list of strings with curly brackets replaced by parentheses.
+.remove_curly_brackets <- function(errors) {
+  errors <- lapply(errors, function(x) gsub("\\{", "(", x))
+  errors <- errors <- lapply(errors, function(x) gsub("\\{", ")", x))
```
Comment (ymarcon) [inline R/datashield.errors.R]: Expecting a closing curly bracket to be replaced instead, right?

File: DESCRIPTION [hunk 6]
```diff
@@ -23,9 +23,13 @@ Depends:
...
     methods,
     progress,
-    R6
+    R6,
+Imports: 
+    cli
 Suggests: 
-    testthat (>= 2.1.0)
+    testthat (>= 2.1.0),
+    DSLite,
```
Comment (ymarcon) [inline DESCRIPTION]: That makes a circular dependency and the CRAN will not accept that.

**Review by ymarcon — CHANGES_REQUESTED** [review]
You are introducing a circular dependency by importing DSLite. You could move the tests to the DSLite package instead?

File: DESCRIPTION (hunk 6 above)
Comment (timcadman) [reply DESCRIPTION]: I think so long as it is in the `suggests` and not the `depends` field it should be ok? But I can remove the tests if you prefer.

Comment (ymarcon) [reply DESCRIPTION]: Yes please move the tests to DSLite.

**Review by ymarcon — CHANGES_REQUESTED** [review]

File: DESCRIPTION [hunk 7]
```diff
@@ -23,9 +23,14 @@ Depends:
...
     progress,
-    R6
+    R6,
+Imports: 
+    cli
 Suggests: 
-    testthat (>= 2.1.0)
+    testthat (>= 2.1.0),
+    DSLite, 
+    dsBase,
```
Comment (ymarcon) [inline DESCRIPTION]: dsBase is not in the CRAN, do not depend on this for your tests;

**Review by ymarcon — CHANGES_REQUESTED** [review]

**Pull request merged**